src/Typedefs.idr

```diff
@@ -3,34 +3,129 @@ module Typedefs
 import Data.Fin
 import Data.Vect
 
-import Types
+import Names
 
 %default total
 %access public export
 
-||| Type definition
-||| @n The number of type variables in the type
-data TDef : (n:Nat) -> Type where
-  ||| The empty type
-  T0    :                                      TDef n
+mutual
+  ||| Type definition
+  ||| @n The number of type variables in the type
+  data TDef : (n:Nat) -> Type where
+    ||| The empty type
+    T0    :                                          TDef n
+
+    ||| The unit type
+    T1    :                                          TDef n
+
+    ||| The coproduct type
+    TSum  :             Vect (2 + k) (TDef n)     -> TDef n
+
+    ||| The product type
+    TProd :             Vect (2 + k) (TDef n)     -> TDef n
+
+    ||| A type variable
+    ||| @i De Bruijn index
+    TVar  :             (i:Fin (S n))             -> TDef (S n)
+
+    ||| Mu
+    TMu   :             Vect k (Name, TDef (S n)) -> TDef n
 
-  ||| The unit type
-  T1    :                                      TDef n
+    ||| Application of a named type to a vector of arguments.
+    TApp  : TNamed k -> Vect k (TDef n)           -> TDef n
 
-  ||| The coproduct type
-  TSum  :         Vect (2 + k) (TDef n)     -> TDef n
+  ||| Named type definition.
+  ||| @n The number of type variables in the type.
+  data TNamed : (n : Nat) -> Type where
```

Comment:
This looks like it can be made a record, packaging `name` and `def` getters inside.